Use ArgumentSpecValidator in AnsibleModule

samdoran opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Switch to using new `ArgumentSpecValidator` class in `AnsibleModule` instead of existing methods in `AnsibleModule`. Remove all private methods in `AnsibleModule` related to argument spec validation.
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`lib/ansible/module_utils/basic.py`
`lib/ansible/module_utils/common/arg_spec.py`
`lib/ansible/module_utils/common/parameters.py`
`lib/ansible/module_utils/common/validation.py`
`lib/ansible/module_utils/errors.py`
`lib/ansible/plugins/action/validate_argument_spec.py`

samdoran committed:
Begin using ArgumentSpecValidator in AnsibleModule

samdoran committed:
Add check parameters to ArgumentSpecValidator

Add additional parameters for specifying required and mutually exclusive parameters.
Add code to the .validate() method that runs these additional checks.

samdoran committed:
Make errors related to unsupported parameters match existing behavior

Update the punctuation in the message slightly to make it more readable.
Add a property to ArgumentSpecValidator to hold valid parameter names.

samdoran committed:
Set default values after performining checks

samdoran committed:
FIx sanity test failure

samdoran committed:
Use correct parameters when checking sub options

samdoran committed:
Use a dict when iterating over check functions

Referencing by key names makes things a bit more readable IMO.

samdoran committed:
Fix bug in comparison for sub options evaluation

samdoran committed:
Add options_context to check functions

This allows the parent parameter to be added the the error message if a validation
error occurs in a sub option.

samdoran committed:
Fix bug in apply_defaults behavior of sub spec validation

samdoran committed:
Accept options_conext in get_unsupported_parameters()

If options_context is supplied, a tuple of parent key names of unsupported parameter will be
created. This allows the full "path" to the unsupported parameter to be reported.

samdoran committed:
Build path to the unsupported parameter for error messages.

samdoran committed:
Remove unused import

samdoran committed:
Update recursive finder test

samdoran committed:
Skip if running in check mode

This was done in the _check_arguments() method. That was moved to a function that has no
way of calling fail_json(), so it must be done outside of validation.

This is a silght change in behavior, but I believe the correct one.

Previously, only unsupported parameters would cause a failure. All other checks would not be executed
if the modlue did not support check mode. This would hide validation failures in check mode.

samdoran committed:
The great purge

Remove all methods related to argument spec validation from AnsibleModule

sivel commented on lib/ansible/module_utils/common/arg_spec.py:
```diff
@@ -36,13 +42,29 @@
+                 required_if=None,
+                 required_by=None,
+                 _kind='module',
+                 _name=None,
```
I'm on the fence here with `_kind` and `_name`.  I think I'd much rather the caller be responsible for this.  But there are tradeoffs.

A possible solution would be to use `{{ kind }}` and `{{ name }}` in the string formatting in here:

```
self._add_error("Unsupported parameters for ({{name}}) {{kind}}: {0}. Supported parameters include: {1}."
```

and then allow the caller to do something like:

```
self.fail_json(msg=self.validator.error_messages[0].format(kind='module', name=self._name))
```

samdoran replied:
I didn't really like this much either. It doesn't seem like information that is relevant to the validator class and is only used in warnings. I mainly did this to appease the existing unit tests which set `AnsibleModule._name` and look for it in the output string.

samdoran replied:
I'll play around with the format string option. That seems like a good way to keep the context in the caller.

samdoran replied:
This does work, and I like that it keeps those bits of information out of the validator class. But it does mean the caller could end up with `Unsupported parameters for ({name}) {kind}:` in their error messages if they don't replace those values. That seems a bit awkward.

Shrews replied:
I would suggest that this validator class not depend on knowing those caller specifics at all, nor put the responsibility of filling in templated error messages on the caller. How about, instead, the `error_messages` attribute be generic lines of text without that data, and we offer a new `error_json` attribute that the caller could parse to generate it's own error messages with any specifics it has?

sivel replied:
I've been thinking the same thing.  Ideally it would be great if the validator raised exceptions, instead of passing back formatted errors, but I get the reasons why it is like it is currently.

Shrews replied:
I explored what it would take to do this, and it's not as straight forward as I'd hoped it would be. Much of the validation actually happens in `parameters.py`, so being able to categorize errors into a json structure in the validator class in `arg_spec.py` would be difficult. It could be done, just would need some slight re-engineering. While I think that would be a more elegant and clean solution to this, I don't think time is going to allow for it and we should probably just stick with what Sam has here now.

FWIW, I think it would be possible to implement the structured data solution later, while also keeping the current solution, so we could provide a better way forward for handling errors and at the same time not break anything currently using the templated message behavior.

samdoran replied:
After talking to @Shrews, I have an idea how to add an interface for raising exceptions vs. returning the formatted messages. But I don't think it will be done in time for release.

samdoran replied:
Unsupported parameters are the only type of failure that requires more context than can be derived from the argument spec and parameters. Maybe we could raise an exception just for unsupported parameters? Or provide a public `unsupported_parameters` attribute the caller needs to handle? Neither are ideal since it makes the interface more complicated, but it would eliminate this problem.

samdoran committed:
Keep _name and kind in the caller and out of the validator

This seems a bit awkward since this means the caller could end up with {name} and {kind} in
the error message if they don't run the messages through the .format() method
with name and kind parameters.

samdoran committed:
Double moustaches work

I wasn't sure if they get stripped or not. Looks like they do. Neat trick.

samdoran committed:
Add changelog

samdoran committed:
Update unsupported parameter test

The error message changed to include name and kind.

samdoran committed:
Remove unused import

samdoran committed:
Add better documentation for ArgumentSpecValidator class

Shrews commented on lib/ansible/module_utils/common/arg_spec.py:
```diff
@@ -27,22 +26,76 @@
+    :param required_together: List of lists of terms that are required together.
+    :type: list, optional
+
+    :param required_one_of: List of lists of terms, on of which in each list
```
s/on/one/

Shrews commented on lib/ansible/module_utils/common/arg_spec.py:
```diff
@@ -27,22 +26,76 @@
+    :type: list, optional
+
+    :param required_if: List of lists of ``[parameter, value, [parameters]]`` where
+        one of [parameters] is required is ``parameter`` == ``value``.
```
s/is/if/

Shrews commented on lib/ansible/module_utils/common/arg_spec.py:
```diff
@@ -27,22 +26,76 @@
+        self.required_together = required_together
+        self.required_one_of = required_one_of
+        self.required_if = required_if
+        self.required_by = required_by
```
These don't appear to be "public" attributes, so we should stay consistent with indicating their privateness by prepending their names with an `_` character.

samdoran committed:
Fix example

samdoran committed:
Few more docs fixes

samdoran committed:
Mark required and mutually exclusive attributes as private

nitzmahone reviewed (commented):
Finally got around to playing with this a bit. I think this checks the major boxes pretty well, but there are a few big-picture API usability things that stick out to me that shouldn't be too hard to address:
* It seems odd that the constructor accepts both the argspec and the args. The `validate` method should probably be the one that accepts the args to be validated, giving you a "validation factory", which you then pass the args to and inspect a result from (possibly with a couple different versions of validate, one of which throws exceptions?)
* That split somewhat implies that the intermediate storage for the in-flight validation and the final results also probably don't belong on the validator object itself, but rather a `ValidationResult` object returned from it.
* +1000 for the exhaustive validation (vs "fail on first error"), though IMO it should be storing the intermediate validation errors as structured data (eg a new `AnsibleValidationError` type hierarchy), then either presenting those raw on a `ValidationResult` object, having some handy utility methods to format/translate the structured errors for display, or wrapping them in a multiple failure exception (depending on the needs of the caller). That would also allow the things like conditional formatting for test compatibility (assuming we think that's important enough) to happen externally.
* Needs a little more thought on what's `_` prefixed, especially with the public validation methods in `validation.py`.

samdoran commented:
> It seems odd that the constructor accepts both the argspec and the args. The validate method should probably be the one that accepts the args to be validated, giving you a "validation factory", which you then pass the args to and inspect a result from (possibly with a couple different versions of validate, one of which throws exceptions?)

This is great feedback. I originally had the `validate()` method accepting the spec and params, but then changed it to be an instantiation parameter. The idea of a validation factory makes sense. I'll work on changing this. I am planning to add another validation method that raises exceptions.

> That split somewhat implies that the intermediate storage for the in-flight validation and the final results also probably don't belong on the validator object itself, but rather a ValidationResult object returned from it.

Would the `AnsibleValidatoError` accept parameters and have attributes that would provide more information beyond an error message? I could use some guidance there. A new exception subclassed from existing exception classes that just contains the same error messages doesn't seem to be adding much value.

> +1000 for the exhaustive validation (vs "fail on first error"), though IMO it should be storing the intermediate validation errors as structured data (eg a new AnsibleValidationError type hierarchy)...

Originally I was collecting the exceptions but wasn't clear on a useful way to store and preset that information. For role arg spec, we just needed the various failure messages, so I ended up returning the list of errors.

>  ...then either presenting those raw on a ValidationResult object, having some handy utility methods to format/translate the structured errors for display, or wrapping them in a multiple failure exception (depending on the needs of the caller).

I think I understand most of this. What would the structured errors look like? A dictionary of failures with parameters and an exception object for each failure? Catching mulitple exceptions I'm familiar with, but not raising multiple exceptions.

> Needs a little more thought on what's _ prefixed, especially with the public validation methods in validation.py.

Agreed. A lot of those probably should not be public, but I wanted feedback from others. I will work on renaming them.

Shrews commented:
> Finally got around to playing with this a bit. I think this checks the major boxes pretty well, but there are a few big-picture API usability things that stick out to me that shouldn't be too hard to address:
> 
> * It seems odd that the constructor accepts both the argspec and the args. The `validate` method should probably be the one that accepts the args to be validated, giving you a "validation factory", which you then pass the args to and inspect a result from (possibly with a couple different versions of validate, one of which throws exceptions?)

If we want to have a single validator validate many different specs, yes, that's the way we should go. Sam and I discussed this a bit early on and this is the way we went, but this could easily be changed.

FWIW, I'm opposed to having an API where we have versions of methods that raise exceptions and alternate versions that do not. It complicates the interface, and the internals will necessarily be more complex to account for both. But this is just my opinion, so take it for what it's worth.


> * That split somewhat implies that the intermediate storage for the in-flight validation and the final results also probably don't belong on the validator object itself, but rather a `ValidationResult` object returned from it.

We discussed this early on as well. This is the final interface I'd like to see in the end, too. Time constraints led us to just having error strings, but we did leave room to implement custom error objects by creating accessor methods to the error results (e.g., `error_messages`). New accessors methods/attributes could return the custom error objects.


> * +1000 for the exhaustive validation (vs "fail on first error"), though IMO it should be storing the intermediate validation errors as structured data (eg a new `AnsibleValidationError` type hierarchy), then either presenting those raw on a `ValidationResult` object, having some handy utility methods to format/translate the structured errors for display, or wrapping them in a multiple failure exception (depending on the needs of the caller). That would also allow the things like conditional formatting for test compatibility (assuming we think that's important enough) to happen externally.

Agreed, I prefer exhaustive validation too. I personally don't see the need to do iterative validation, but there is a path to that, if we decide we want it.


> * Needs a little more thought on what's `_` prefixed, especially with the public validation methods in `validation.py`.

samdoran commented:
I'm getting a clearer idea of what to do here. Rather than multiple methods with different behavior, `ArgumentSpecValidator.validate(argument_spec, parameters)` will return the validated parameters in a `ValidationResult` object, or it will raise an `AnsibleValidationException` if there are any failures. The `errors` property of `AnsibleValidationException` will contain all the exceptions.

Maybe something like this:

```python
validator = ArgumentSpecValidator(argument_spec)
try:
    result = validator.validate(parameters)
except ValidationMultipleError:
    for e in ValidationMultipleError.errors:
        if isinstance(e, TypeError):
            print('Type failure: %s' % to_native(e))
        elif isinstance(e, ValueError):
            print("Value failure: %s" % to_native(e))
        else:
            print(to_native(e))
params = result.validated_parameters
print(params['foo'])
```

samdoran committed:
Mark validate functions as private

samdoran committed:
Reorganize functions in validation.py

samdoran committed:
Remove unused imports in basic.py related to argument spec validation

samdoran committed:
Create errors is module_utils

We have errors in lib/ansible/errors/ but those cannot be used by modules.

samdoran committed:
Update recursive finder test

samdoran committed:
Move errors to file rather than __init__.py

samdoran committed:
Change ArgumentSpecValidator.validate() interface

Raise AnsibleValidationErrorMultiple on validation error which contains all AnsibleValidationError
exceptions for validation failures.

Return the validated parameters if validation is successful rather than True/False.

Update docs and tests.

samdoran committed:
Get attribute in loop so that the attribute name can also be used as a parameter

samdoran committed:
Shorten line

samdoran committed:
Update calling code in AnsibleModule for new validator interface

samdoran committed:
Update calling code in validate_argument_spec based in new validation interface

samdoran commented:
I have made a good bit of progress on implementing custom exceptions for validation failure and change the interface to raise exceptions or return the validate result.

I still need to change the `validate()` interface to accept the parameters.

samdoran committed:
Base custom exception class off of Exception

samdoran committed:
Call the __init__ method of the base Exception class to populate args

samdoran committed:
Ensure no_log values are always updated

samdoran committed:
Make custom exceptions more hierarchical

This redefines AnsibleError from lib/ansible/errors with a different signature since that cannot
be used by modules. This may be a bad idea. Maybe lib/ansible/errors should be moved to
module_utils, or AnsibleError defined in this commit should use the same signature as the original.

samdoran committed:
Just go back to basing off Exception

samdoran committed:
Return ValidationResult object on successful validation

Create a ValidationResult class.
Return a ValidationResult from ArgumentSpecValidator.validate() when validation is successful.
Update class and method docs.
Update unit tests based on interface change.

samdoran committed:
Make it easier to get error objects from AnsibleValidationResultMultiple

This makes the interface cleaner when getting individual error objects contained in a single
AnsibleValidationResultMultiple instance.

samdoran committed:
Define custom exception for each type of validation failure

These errors indicate where a validation error occured. Currently they are empty but could
contain specific data for each exception type in the future.

samdoran committed:
Update tests based on (yet another) interface change

samdoran committed:
Mark several more functions as private

These are all doing rather "internal" things. The ArgumentSpecValidator class is the preferred
public interface.

samdoran committed:
Move warnings and deprecations to result object

Rather than calling deprecate() and warn() directly, store them on the result object so the
caller can decide what to do with them.

samdoran committed:
Use subclass for module arg spec validation

The subclass uses global warning and deprecations feature

samdoran commented:
Based on feedback I have made the following changes:
- `ArgumentSpecValidator.validate()` always returns a `ValidationResult` object
- Created a custom Exception for each type of validation failure, all subclassed from `AnsibleValidationError`
- Created `ModuleArgumentSpecValidator` subclass that calls `warn()` and `deprecate()`
- Moved no_log values and error messages to the `ValidationResult` object

The various exception names could probably be bikeshed to death. They are all currently empty objects but could accommodate holding structured data in the future.

This  changes the main interface to require the caller to inspect the `ValidationResult` object to determine if validation was successful or not. While it would be more idiomatic to raise an exception on failure, it complicated the interface.

In the event of validation failure, no `ValidationResult` object was returned. This meant the caller had no way to get no_log values, deprecations, or warnings (I could smuggle that out in the raised `AnsibleValidationFailureMultipl` object, but that feels wrong). It also created tension between where to store that data so it was always accessible. Storing it on the `ArgumentSpecValidator` instance did not make sense because that should be a "validation factory" — it should not be storing any state related to the validation.

sivel commented on lib/ansible/module_utils/common/arg_spec.py:
```diff
@@ -5,130 +5,266 @@
+    RequiredIfError,
+    RequiredOneOfError,
+    RequiredTogetherError,
+    UnsupportedError,
```
I love having all of these discrete exceptions for errors.

sivel commented on lib/ansible/module_utils/common/arg_spec.py:
```diff
@@ -5,130 +5,266 @@
 
         for option, alias in alias_warnings:
-            warn('Both option %s and its alias %s are set.' % (option, alias))
+            result.warnings.append({'option': option, 'alias': alias})
```
I notice here and in `ModuleArgumentSpecValidator` that `result.warnings` seems to be specifically related to "alias warnings".  Should we differentiate that?  Or should we make it more generic?

samdoran replied:
It should probably be more generic. It just so happens that the only warning from the validation code is related to aliases, but it would be helpful to have it be more generic.

samdoran replied:
Maybe warnings and deprecations should be objects not just dicts? I didn't have a clear vision of how best to structure that data so it could be easily used later, so I just started with a basic dict.

sivel replied:
I'm on the fence about holding this up to try and come up with something clever that seems like we know what the future will hold.  With this being the only use case now, it's hard to know what another warning could look like in the future, and how we might need to represent that.

nitzmahone replied:
Maybe just `_` prefix `warnings`on `ValidationResult` for now? This definitely doesn't seem like it'd be generally useful as-is, and we'd almost certainly have to restructure it later to create actual warning objects if we have more than one (rather than a list of dicts).

samdoran committed:
Fix up docs

samdoran committed:
Remove legal_inputs munging from _handle_aliases()

This is done in AnsibleModule by the _set_internal_properties() method. It only makes sense
to do that for an AnsibleModule instance (it should update the parameters before performing
validation) and shouldn't be done by the validator.

Create a private function just for getting legal inputs since that is done in a couple of places.

It may make sense store that on the ValidationResult object.

samdoran committed:
Increase test coverage

samdoran commented:
bot_skip

samdoran committed:
Remove unnecessary conditional

ci_complete

nitzmahone commented on lib/ansible/module_utils/common/arg_spec.py:
```diff
@@ -5,130 +5,265 @@
+        self.argument_spec = argument_spec
+
+    @property
+    def valid_parameter_names(self):
```
This property should probably also be `_` prefixed- the way the aliases are encoded in here right now makes it pretty useless for anything but generating that string in the error case.

samdoran replied:
I agree with the above three statements. I _really_ dislike the duplication of sub-spec validation in its own explicit step, but refactoring that was too much while moving all the other validation code. I'll put a FIXME in there since it really should be integrated into the existing validation rather than its own thing.

> we'd almost certainly have to restructure it later to create actual warning objects if we have more than one

I was thinking custom objects would make sense as well. It's unfortunate that the primary use case right now is generating an error message string, so that's clouding my thinking.

This is more or less the same problem with `valid_parameter_names` as well. But I need some time to think through how to make this more generally useful. I will mark these as private for now.

nitzmahone commented on lib/ansible/module_utils/common/parameters.py:
```diff
@@ -799,53 +722,166 @@ def validate_sub_spec(argument_spec, parameters, prefix='', options_context=None
-                    (check_required_if, 'required_if'),
-                    (check_required_by, 'required_by'),
-                ]
+                checks = (
```
this seems like it should be shared with the main one rather than duplicated here

samdoran replied:
This has tripped me up a few times already: I'll update the code in the `validate()` method but forget to do it in this function. Yet another argument for combining the sub-spec validation. I'll make a variable they can share for now.

nitzmahone reviewed (changes requested):
Probably just a couple other things we should `_` prefix- I really wish we could get the subspec validation and the outer spec validation impls shared, but at least things like the validator dispatch table should be pretty easy to share. I'm good to merge for 2.11 once those are in (and you could probably successfully fight me on the dispatch table one too ;) )

samdoran committed:
Mark warnings and deprecations as private in the ValidationResult

They can be made public once we come up with a way to make them more generally useful,
probably by creating cusom objects to store the data in more structure way.

samdoran committed:
Mark valid_parameter_names as private and populate it during initialization

samdoran committed:
Use a global for storing the list of additonal checks to perform

This list is used by the main validate method as well as the sub spec validation.

felixfontein commented:
This PR breaks some things in collections, I've created #73983 for tracking that.
